# kmer_decomposition

## Workflows
### <a href="https://github.com/CobiontID/kmer_decomposition/tree/main/readviz_pipeline">Read k-mer decomposition and visualisation</a>

Tallies k-mers in a read set, reduce to two dimensions and visualise read clusters (defaults to tetranucleotides). Annotates read plots with additional sequence features, such as estimated coding density, approximate k-mer coverage and sequence k-mer diversity.

#### Example data set: _Erannis defoliaria_
Decomposed read tetranucleotides from _Erannis defoliaria_ indicate the presence of bacteria in the sample (top). The reads are coloured by estimated coding density.

<img src="https://github.com/CobiontID/kmer_decomposition/blob/main/ilEraDefo1_hexamer.2d_plot_labelled.png" width=500>

### <a href="https://github.com/CobiontID/kmer_decomposition/tree/main/contigviz_pipeline">Contig/scaffold k-mer decomposition and visualisation</a>

As with the reads, tallies and reduces tetranucleotide composition to two dimensions and plots with annotations. In addition to estimated coding density and k-mer diversity, FastK provides a measure of repetitiveness, and coverage for primary Hifiasm assemblies can be extracted and used to annotate the plots. A selection tool allows sequences that are of interest to be selected and downloaded with their annotations. Where Hi-C data are available, a SALSA pair file may also be provided to annotate plots with scaffold connectivity information.

#### Example data set: Hylocomiadelphus triquetrus
![image](https://user-images.githubusercontent.com/10507101/133108115-a3dbe6af-a602-47d9-a56c-27887d464084.png)


## Tools
### <a href="https://github.com/CobiontID/kmer_decomposition/tree/main/kmer-counter">kmer-counter</a>
Counts the number of occurences of each k-mer of size k for each record in a fasta file of nucleotide sequences (canonicalised or non-canonicalised). Implemented in Rust, runs approximately ten times faster than the equivalent code in Python.
### <a href="https://github.com/CobiontID/kmer_decomposition/tree/main/unique-kmer-counts">unique-kmer-counter</a>
Count the number of distinct k-mers of size k for each record in a fasta files of nucleotide sequences, and divide by sequence length. Implemented in Rust.
### hexamer
Estimates the coding density using the sum of lengths of putative coding sequences divided by sequence length. The cobiont pipelines previously used a modified version of the old hexamer code. The relevant functionality is now available in an updated version of hexamer from
https://github.com/richarddurbin/hexamer (to extract the estimated density, pipe stdout to `awk '{ print $3/$2}'`)

### FastK profile medians

### <a href="https://github.com/CobiontID/kmer_decomposition/tree/main/VAE">Variational Autoencoder for k-mer decomposition</a>
#### vae.py
Read k-mer counts are reduced to two dimensions following the method of <a href="https://arxiv.org/abs/1312.6114">Kingma and Welling (2013)</a>. Outputs two-dimensional representation of the read set and a basic plot.
#### vae_draw.py
Renders two-dimensional representations generated by vae.py with optional coloured annotation
#### labelling utils
- `category_labels.py` Assign sequences to bins based on whether they are contained in a set defined by a file (e.g. lists of sequences identified as belonging to a particular taxon)
- `category_labels_from_cont.py`Assign sequences to bins based on a continuous vector of numbers, either using quantiles (default) or linear slices (e.g. estimated coding density)
- `category_labels_from_cont_manual.py` Assign sequences to bins based on a list of boundaries (e.g. to slice k-mer coverage estimates at the minima of a histogram)

### <a href="https://github.com/CobiontID/kmer_decomposition/blob/main/draw_contigs/">Draw contigs/scaffolds</a>
**Select_contigs_reduced_multi.py**

Generate HTML file of decomposed tetranucleotide plots with binned annotations.

### <a href="https://github.com/CobiontID/kmer_decomposition/blob/main/Hi-C/utils/">Hi-C</a>
- `hic_links.py` Generate scaffold connectivity matrix from SALSA pair file and store the resulting np array in a .npy file. Optionally returns a 1/0 vector for connected/uncnnected scaffolds, or (number of connections)/(scaffold length), for use with `Select_contigs_reduced_multi.py`.
